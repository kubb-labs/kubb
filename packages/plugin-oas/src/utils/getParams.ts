import { isParameterObject } from '@kubb/oas'

import { camelCase } from '@kubb/core/transformers'
import type { FunctionParamsAST } from '@kubb/core/utils'
import type { OasTypes } from '@kubb/oas'
import type { Params } from '@kubb/react/types'
import type { OperationSchema } from '../types.ts'
/**
 *
 * @deprecated
 * TODO move to operationManager hook
 */
export function getASTParams(
  operationSchema: OperationSchema | undefined,
  {
    typed = false,
    override,
  }: {
    typed?: boolean
    override?: (data: FunctionParamsAST) => FunctionParamsAST
  } = {},
): FunctionParamsAST[] {
  if (!operationSchema || !operationSchema.schema.properties || !operationSchema.name) {
    return []
  }

  return Object.entries(operationSchema.schema.properties).map(([name, schema]: [string, OasTypes.SchemaObject]) => {
    const isParam = isParameterObject(schema)
    const data: FunctionParamsAST = {
      name,
      enabled: !!name,
      required: isParam ? schema.required : true,
      type: typed ? `${operationSchema.name}["${name}"]` : undefined,
    }

    return override ? override(data) : data
  })
}

export function getPathParams(
  operationSchema: OperationSchema | undefined,
  options: {
    typed?: boolean
    override?: (data: FunctionParamsAST) => FunctionParamsAST
  } = {},
) {
  return getASTParams(operationSchema, options).reduce((acc, curr) => {
    if (curr.name && curr.enabled) {
      acc[camelCase(curr.name)] = {
        default: curr.default,
        type: curr.type,
        optional: !curr.required,
      }
    }

    return acc
  }, {} as Params)
}
