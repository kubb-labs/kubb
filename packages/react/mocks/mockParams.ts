import type { Params } from '../src/types.ts'

export const mockParams: Array<{ name: string; params: Params }> = [
  {
    name: 'empty',
    params: {
      name: {},
    },
  },
  {
    name: 'string',
    params: {
      name: {
        type: 'string',
      },
    },
  },
  {
    name: 'optional string',
    params: {
      name: {
        type: 'string',
        optional: true,
      },
    },
  },
  {
    name: 'default string',
    params: {
      name: {
        type: 'string',
        default: 'test',
      },
    },
  },
  {
    name: 'undefined',
    params: {
      name: undefined,
    },
  },
  {
    name: 'name and address',
    params: {
      name: {
        type: 'string',
      },
      address: {
        type: 'string',
      },
    },
  },
  {
    name: 'name[optional], age[optional] and address',
    params: {
      name: {
        type: 'string',
        optional: true,
      },
      age: {
        type: 'number',
        optional: true,
      },
      address: {
        type: 'string',
      },
    },
  },
  {
    name: 'name[optional] age, address[optional] and country[Belgium]',
    params: {
      name: {
        type: 'string',
        optional: true,
      },
      age: {
        type: 'number',
      },
      address: {
        type: 'string',
        optional: true,
      },
      country: {
        type: 'string',
        default: '"Belgium"',
      },
    },
  },
  {
    name: 'params[with name] with age, address[optional] and country',
    params: {
      data: {
        type: 'Data',
        children: {
          age: {
            type: 'number',
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
          },
        },
      },
    },
  },
  {
    name: 'params[without name] with age, address[optional] and country',
    params: {
      data: {
        children: {
          age: {
            type: 'number',
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
          },
        },
      },
    },
  },
  {
    name: 'undefined children param',
    params: {
      data: {
        children: {
          age: undefined,
        },
      },
    },
  },
  {
    name: 'empty children param',
    params: {
      data: {
        children: {
          age: {},
        },
      },
    },
  },
  {
    name: 'params[without name and default] with age[optional, 90] address[optional] and country[optional, Belgium]',
    params: {
      data: {
        default: '{}',
        children: {
          age: {
            type: 'number',
            default: '90',
            optional: true,
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
            default: "'Belgium'",
            optional: true,
          },
        },
      },
    },
  },
  {
    name: 'params[without name] with age[optional, 20] address[optional] and country[optional, Belgium]',
    params: {
      data: {
        children: {
          age: {
            type: 'number',
            default: '20',
            optional: true,
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
            default: '"Belgium"',
            optional: true,
          },
        },
      },
    },
  },
  {
    name: 'params[without name and mode inline] with age[optional, 20] address[optional] and country[optional, Belgium]',
    params: {
      data: {
        mode: 'inline',
        children: {
          age: {
            type: 'number',
            default: '20',
            optional: true,
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
            default: '"Belgium"',
            optional: true,
          },
        },
      },
    },
  },
  {
    name: 'params[without name and mode inline] with age[20] address and country[Belgium]',
    params: {
      data: {
        mode: 'inline',
        children: {
          age: {
            type: 'number',
            default: '20',
          },
          address: {
            type: 'string',
          },
          country: {
            type: 'string',
            default: '"Belgium"',
          },
        },
      },
    },
  },
  {
    name: 'params[without name and mode inlineSpread] with age[20] address and country[Belgium]',
    params: {
      data: {
        mode: 'inlineSpread',
        children: {
          age: {
            type: 'number',
            default: '20',
          },
          address: {
            type: 'string',
          },
          country: {
            type: 'string',
            default: '"Belgium"',
          },
        },
      },
    },
  },
]
