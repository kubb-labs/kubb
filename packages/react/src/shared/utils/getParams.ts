import { camelCase } from '@kubb/core/transformers'
import { orderBy } from 'natural-orderby'

export type Param = {
  /**
   * `object` will return the pathParams as an object.
   *
   * `inline` will return the pathParams as comma separated params.
   * @default `'inline'`
   * @private
   */
  mode?: 'object' | 'inline' | 'inlineSpread'
  type?: 'string' | 'number' | (string & {})
  optional?: boolean
  /**
   * @example test = "default"
   */
  default?: string
  /**
   * Used for no TypeScript(with mode object)
   * @example test: "default"
   */
  value?: string
  children?: Params
}

type ParamItem =
  | (Pick<Param, 'mode' | 'type' | 'value'> & {
      optional?: true
      default?: never
      children?: Params
    })
  | (Pick<Param, 'mode' | 'type' | 'value'> & {
      optional?: false
      default?: string
      children?: Params
    })

export type Params = Record<string, Param | undefined>

type Options = {
  type: 'constructor' | 'call'
}

function order(items: Array<[key: string, item?: ParamItem]>) {
  return orderBy(
    items.filter(Boolean),
    [
      ([_key, item]) => {
        if (item?.children) {
          return undefined
        }
        return !item?.default
      },
      ([_key, item]) => {
        if (item?.children) {
          return undefined
        }
        return !item?.optional
      },
    ],
    ['desc', 'desc'],
  )
}

function parseChild(key: string, item: ParamItem, options: Options): string[] {
  const entries = order(Object.entries(item.children as ParamItem))

  const types: string[] = []
  const names: string[] = []

  const optional = entries.every(([_key, item]) => item?.optional)

  entries.forEach(([key, entryItem]) => {
    if (entryItem) {
      if (options.type === 'call') {
        names.push(...parseItem(camelCase(key), { ...entryItem, type: undefined }))
      } else {
        names.push(
          ...parseItem(camelCase(key), {
            ...entryItem,
            type: undefined,
            value: undefined,
          }),
        )
      }

      if (entries.some(([_key, item]) => item?.type)) {
        types.push(...parseItem(camelCase(key), { ...entryItem, default: undefined }))
      }
    }
  })

  const name = item.mode === 'inline' ? key : names.length ? `{ ${names.join(', ')} }` : ''

  const type = item.type ? item.type : types.length ? `{ ${types.join('; ')} }` : undefined

  return parseItem(name, {
    type: options.type === 'constructor' ? type : undefined,
    default: item.default ? item.default : undefined,
    optional: !item.default ? optional : undefined,
  } as ParamItem)
}

function parseItem(name: string, item: ParamItem): string[] {
  const acc = []

  if (item.type) {
    if (item.optional) {
      acc.push(`${name}?: ${item.type}`)
    } else {
      acc.push(`${name}: ${item.type}${item.default ? ` = ${item.default}` : ''}`)
    }
  } else if (item.default) {
    acc.push(`${name} = ${item.default}`)
  } else if (item.value) {
    acc.push(`${name} : ${item.value}`)
  } else if (item.mode === 'inlineSpread') {
    acc.push(`... ${name}`)
  } else {
    acc.push(name)
  }

  return acc
}

export function getParams(items: Params, options: Options): string {
  const entries = order(Object.entries(items as Record<string, ParamItem | undefined>))

  return entries
    .reduce((acc, [key, item]) => {
      if (!item) {
        return acc
      }

      if (item.children) {
        if (Object.keys(item.children).length === 0) {
          return acc
        }

        if (item.mode === 'inlineSpread') {
          return [...acc, getParams(item.children!, options)]
        }

        const parsedItem = parseChild(key, item, options)

        return [...acc, ...parsedItem]
      }

      const parsedItem = parseItem(camelCase(key), item)

      return [...acc, ...parsedItem]
    }, [] as string[])
    .join(', ')
}

export function isParamItems(items: any): items is Params {
  return typeof items !== 'string' && items && Object.keys(items)?.length
}
