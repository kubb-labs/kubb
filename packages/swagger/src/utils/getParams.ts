import transformers from '@kubb/core/transformers'
import { FunctionParams } from '@kubb/core/utils'

import { isParameterObject } from './isParameterObject.ts'

import type { FunctionParamsAST } from '@kubb/core/utils'
import type { OasTypes } from '../oas/index.ts'
import type { OperationSchema } from '../types.ts'

export function getASTParams(
  operationSchema: OperationSchema | undefined,
  {
    typed = false,
    override,
    asObject = false,
  }: {
    typed?: boolean
    asObject?: boolean
    override?: (data: FunctionParamsAST) => FunctionParamsAST
  } = {},
): FunctionParamsAST[] {
  if (!operationSchema || !operationSchema.schema.properties || !operationSchema.name) {
    return []
  }

  if (asObject) {
    const nameText = getASTParams(operationSchema)
      .map((item) => (item.name ? transformers.camelCase(item.name) : item.name))
      .join(', ')

    return [
      {
        name: `{ ${nameText} }`,
        type: operationSchema?.name,
        enabled: !!operationSchema?.name,
        required: true,
      },
    ]
  }

  return Object.entries(operationSchema.schema.properties).map(([name, schema]: [string, OasTypes.SchemaObject]) => {
    const isParam = isParameterObject(schema)
    const data: FunctionParamsAST = {
      name,
      required: isParam ? schema.required : undefined,
      type: typed ? `${operationSchema.name}["${name}"]` : undefined,
    }

    return override ? override(data) : data
  })
}

type GetParamsResult = {
  ast: FunctionParamsAST[]
  toString: () => string
}
// TODO convert to class together with `createFunctionParams` and `getASTParams`
export function getParams(
  operationSchema: OperationSchema | undefined,
  {
    typed = false,
    override,
    asObject = false,
  }: {
    typed?: boolean
    asObject?: boolean
    override?: (data: FunctionParamsAST) => FunctionParamsAST
  } = {},
): GetParamsResult {
  const ast = getASTParams(operationSchema, { typed, override, asObject })
  const functionParams = new FunctionParams()
  functionParams.add(ast)

  return {
    ast,
    toString: () => functionParams.toString(),
  }
}
