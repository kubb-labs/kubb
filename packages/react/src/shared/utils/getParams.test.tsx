import { getParams } from './getParams.ts'
import { mockParams } from '../../../mocks/mockParams.ts'

describe('getParams callable(Function.Call)', () => {
  test.each(mockParams)('$name', async ({ name, params }) => {
    expect(getParams(params, { type: 'call' })).toMatchSnapshot()
  })
})

describe('getParams not callable(Function)', () => {
  test.each(mockParams)('$name', async ({ name, params }) => {
    expect(getParams(params, { type: 'constructor' })).toMatchSnapshot()
  })
})
