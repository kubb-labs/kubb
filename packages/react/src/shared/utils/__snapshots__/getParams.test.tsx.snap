// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getParams > 'default string' 1`] = `"name: string = test"`;

exports[`getParams > 'empty children param' 1`] = `"{ age }"`;

exports[`getParams > 'empty' 1`] = `"name"`;

exports[`getParams > 'name and address' 1`] = `"name: string, address: string"`;

exports[`getParams > 'name[optional] age, address[optional]…' 1`] = `"age: number, name?: string, address?: string, country: string = "Belgium""`;

exports[`getParams > 'name[optional], age[optional] and add…' 1`] = `"address: string, name?: string, age?: number"`;

exports[`getParams > 'optional string' 1`] = `"name?: string"`;

exports[`getParams > 'params[with name] with age, address[o…' 1`] = `"{ age, country, address }: Data"`;

exports[`getParams > 'params[without name and default] with…' 1`] = `"{ address, age = 90, country = 'Belgium' }: { address?: string; age?: number; country?: string } = {}"`;

exports[`getParams > 'params[without name and mode inline] …' 1`] = `"data?: { address?: string; age?: number; country?: string }"`;

exports[`getParams > 'params[without name and mode inline] …' 2`] = `"data: { address: string; age: number; country: string }"`;

exports[`getParams > 'params[without name and mode inlineSp…' 1`] = `"address: string, age: number = 20, country: string = "Belgium""`;

exports[`getParams > 'params[without name] with age, addres…' 1`] = `"{ age, country, address }: { age: number; country: string; address?: string }"`;

exports[`getParams > 'params[without name] with age[optiona…' 1`] = `"{ address, age = 20, country = "Belgium" }?: { address?: string; age?: number; country?: string }"`;

exports[`getParams > 'string' 1`] = `"name: string"`;

exports[`getParams > 'undefined children param' 1`] = `""`;

exports[`getParams > 'undefined' 1`] = `""`;
